File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/processor/TokenDecryptionProcessor.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.processor;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.text.ParseException;
import java.util.List;
import java.util.Map;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWEObject;
import com.nimbusds.jose.crypto.DirectDecrypter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.cli.io.TokenReader;
import com.truveta.opentoken.cli.io.TokenWriter;
import com.truveta.opentoken.tokens.Token;
import com.truveta.opentoken.tokentransformer.DecryptTokenTransformer;

/**
 * Process encrypted tokens for decryption.
 * <p>
 * This class is used to read encrypted tokens from input source,
 * decrypt them, and write the decrypted tokens to the output data source.
 */
public final class TokenDecryptionProcessor {

    private static final Logger logger = LoggerFactory.getLogger(TokenDecryptionProcessor.class);
    private static final String V1_TOKEN_PREFIX = "ot.V1.";

    private TokenDecryptionProcessor() {
    }

    /**
     * Reads encrypted tokens from the input data source, decrypts them, and
     * writes the result back to the output data source.
     * 
    * @param reader        TokenReader providing encrypted token rows
    * @param writer        TokenWriter for output
    * @param decryptor     The decryption transformer
    * @param encryptionKey Encryption key used for JWE token decryption
     * @throws IOException if an I/O error occurs
     */
    public static void process(TokenReader reader,
            TokenWriter writer,
            DecryptTokenTransformer decryptor,
            String encryptionKey) throws IOException {
        long rowCounter = 0;
        long decryptedCounter = 0;
        long errorCounter = 0;

        while (reader.hasNext()) {
            Map<String, String> row = reader.next();
            rowCounter++;

            String token = row.get(TokenConstants.TOKEN);

            // Decrypt the token if it's not blank
            if (token != null && !token.isEmpty() && !Token.BLANK.equals(token)) {
                try {
                    String decryptedToken = decryptToken(token, decryptor, encryptionKey);
                    row.put(TokenConstants.TOKEN, decryptedToken);
                    decryptedCounter++;
                } catch (Exception e) {
                    logger.error("Failed to decrypt token for RecordId {}, RuleId {}: {}",
                            row.get(TokenConstants.RECORD_ID), row.get(TokenConstants.RULE_ID), e.getMessage());
                    errorCounter++;
                    // Keep the encrypted token in case of error
                }
            }

            // Write token
            writer.writeToken(row);

            if (rowCounter % 10000 == 0) {
                logger.info(String.format("Processed \"%,d\" tokens", rowCounter));
            }
        }

        logger.info(String.format("Processed a total of %,d tokens", rowCounter));
        logger.info(String.format("Successfully decrypted %,d tokens", decryptedCounter));
        if (errorCounter > 0) {
            logger.warn(String.format("Failed to decrypt %,d tokens", errorCounter));
        }
    }

    private static String decryptToken(String token,
            DecryptTokenTransformer decryptor,
            String encryptionKey) throws Exception {
        if (token.startsWith(V1_TOKEN_PREFIX)) {
            return decryptV1Token(token, decryptor, encryptionKey);
        }
        return decryptor.transform(token);
    }

    private static String decryptV1Token(String token,
            DecryptTokenTransformer decryptor,
            String encryptionKey) throws ParseException, JOSEException {
        if (encryptionKey == null || encryptionKey.isBlank()) {
            throw new IllegalArgumentException("Encryption key is required for JWE token decryption");
        }

        String jweCompact = token.substring(V1_TOKEN_PREFIX.length());
        JWEObject jweObject = JWEObject.parse(jweCompact);
        DirectDecrypter decrypter = new DirectDecrypter(encryptionKey.getBytes(StandardCharsets.UTF_8));
        jweObject.decrypt(decrypter);

        Map<String, Object> payload = jweObject.getPayload().toJSONObject();
        Object ppidValue = null;
        if (payload != null) {
            Object ppid = payload.get("ppid");
            if (ppid instanceof List<?> list && !list.isEmpty()) {
                ppidValue = list.get(0);
            } else {
                ppidValue = ppid;
            }
        }

        if (ppidValue == null) {
            return null;
        }

        String ppidToken = ppidValue.toString();
        if (ppidToken.isBlank() || Token.BLANK.equals(ppidToken)) {
            return ppidToken;
        }

        try {
            return decryptor.transform(ppidToken);
        } catch (Exception e) {
            logger.debug("Failed to decrypt legacy token inside JWE payload", e);
            return ppidToken;
        }
    }
}

File: lib/python/opentoken-cli/src/main/opentoken_cli/processor/token_decryption_processor.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import base64
import json
import logging

from jwcrypto import jwe, jwk
from opentoken_cli.io.token_reader import TokenReader
from opentoken_cli.io.token_writer import TokenWriter
from opentoken_cli.processor.token_constants import TokenConstants
from opentoken.tokens.token import Token
from opentoken.tokentransformer.decrypt_token_transformer import DecryptTokenTransformer
from opentoken.tokentransformer.match_token_constants import PAYLOAD_KEY_PPID
from opentoken.tokentransformer.match_token_constants import V1_TOKEN_PREFIX


logger = logging.getLogger(__name__)


class TokenDecryptionProcessor:
    """
    Process encrypted tokens for decryption.
    
    This class is used to read encrypted tokens from input source,
    decrypt them, and write the decrypted tokens to the output data source.
    """

    @staticmethod
    def process(reader: TokenReader, writer: TokenWriter, decryptor: DecryptTokenTransformer):
        """
        Reads encrypted tokens from the input data source, decrypts them, and
        writes the result back to the output data source.
        
        Args:
            reader: Iterator providing encrypted token rows
            writer: Writer instance with write_token method
            decryptor: The decryption transformer
        """
        TokenDecryptionProcessor.process_with_key(reader, writer, decryptor, None)

    @staticmethod
    def process_with_key(
        reader: TokenReader,
        writer: TokenWriter,
        decryptor: DecryptTokenTransformer,
        encryption_key: str | None,
    ):
        """
        Reads encrypted tokens from the input data source, decrypts them, and
        writes the result back to the output data source.

        Args:
            reader: Iterator providing encrypted token rows
            writer: Writer instance with write_token method
            decryptor: The decryption transformer
            encryption_key: Encryption key for decrypting JWE tokens
        """
        row_counter = 0
        decrypted_counter = 0
        error_counter = 0

        for row in reader:
            row_counter += 1
            
            token = row.get(TokenConstants.TOKEN, '')
            
            # Decrypt the token if it's not blank
            if token and token != Token.BLANK:
                try:
                    decrypted_token = TokenDecryptionProcessor._decrypt_token(
                        token, decryptor, encryption_key
                    )
                    row[TokenConstants.TOKEN] = decrypted_token
                    decrypted_counter += 1
                except Exception as e:
                    logger.error(
                        f"Failed to decrypt token for RecordId {row.get(TokenConstants.RECORD_ID)}, "
                        f"RuleId {row.get(TokenConstants.RULE_ID)}: {e}")
                    error_counter += 1
                    # Keep the encrypted token in case of error
            
            writer.write_token(row)

            if row_counter % 10000 == 0:
                logger.info(f'Processed "{row_counter:,}" tokens')

        logger.info(f'Processed a total of {row_counter:,} tokens')
        logger.info(f'Successfully decrypted {decrypted_counter:,} tokens')
        if error_counter > 0:
            logger.warning(f'Failed to decrypt {error_counter:,} tokens')

    @staticmethod
    def _decrypt_token(
        token: str,
        decryptor: DecryptTokenTransformer,
        encryption_key: str | None,
    ) -> str:
        if token.startswith(V1_TOKEN_PREFIX):
            return TokenDecryptionProcessor._decrypt_v1_token(token, decryptor, encryption_key)
        return decryptor.transform(token)

    @staticmethod
    def _decrypt_v1_token(
        token: str,
        decryptor: DecryptTokenTransformer,
        encryption_key: str | None,
    ) -> str:
        if not encryption_key:
            raise ValueError("Encryption key is required for JWE token decryption")

        jwe_compact = token[len(V1_TOKEN_PREFIX):]
        key_bytes = encryption_key.encode('utf-8')
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode('utf-8').rstrip('=')
        jwk_key = jwk.JWK(kty="oct", k=key_b64)

        jwe_token = jwe.JWE()
        jwe_token.deserialize(jwe_compact)
        jwe_token.decrypt(jwk_key)

        payload = json.loads(jwe_token.payload.decode('utf-8'))
        ppid_value = payload.get(PAYLOAD_KEY_PPID, [])
        if isinstance(ppid_value, list):
            ppid_value = ppid_value[0] if ppid_value else ""

        if not ppid_value or ppid_value == Token.BLANK:
            return ppid_value

        try:
            return decryptor.transform(ppid_value)
        except Exception as e:
            logger.debug("Failed to decrypt legacy token inside JWE payload", exc_info=e)
            return ppid_value
